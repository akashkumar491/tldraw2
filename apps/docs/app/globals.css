@tailwind base;
@tailwind components;
@tailwind utilities;
@import url('./demo.css');

html {
	scroll-padding-top: 72px;
}

@layer base {
	.prose h2 a,
	.prose h3 a,
	.prose h4 a {
		@apply font-bold;
	}
	.prose blockquote p {
		margin: 0;
	}
	.prose blockquote p:before,
	.prose blockquote p:after {
		content: none;
	}
	.prose details.article__table-of-contents {
		display: none !important;
	}
	.prose .hljs,
	.not-prose .hljs {
		background: transparent !important;
	}
	input[type='search']::-webkit-search-decoration,
	input[type='search']::-webkit-search-cancel-button,
	input[type='search']::-webkit-search-results-button,
	input[type='search']::-webkit-search-results-decoration {
		-webkit-appearance: none;
	}

	@media (max-width: 768px) {
		.prose table {
			table-layout: fixed;
		}

		.prose td {
			word-break: break-word;
		}
	}
}
