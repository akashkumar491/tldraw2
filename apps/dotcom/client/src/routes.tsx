import { captureException } from '@sentry/react'
import {
	READ_ONLY_LEGACY_PREFIX,
	READ_ONLY_PREFIX,
	ROOM_PREFIX,
	SNAPSHOT_PREFIX,
} from '@tldraw/dotcom-shared'
import { useEffect } from 'react'
import { Outlet, Route, createRoutesFromElements, useRouteError } from 'react-router-dom'
import { DefaultErrorFallback } from './components/DefaultErrorFallback/DefaultErrorFallback'
import { ErrorPage } from './components/ErrorPage/ErrorPage'

export const router = createRoutesFromElements(
	<Route
		element={
			// Add all top level providers that require the router here

			<Outlet />
		}
		ErrorBoundary={() => {
			const error = useRouteError()
			useEffect(() => {
				captureException(error)
			}, [error])
			return (
				<ErrorPage
					messages={{
						header: 'Something went wrong',
						para1:
							'Please try refreshing the page. Still having trouble? Let us know at hello@tldraw.com.',
					}}
				/>
			)
		}}
	>
		<Route errorElement={<DefaultErrorFallback />}>
			<Route path="/" lazy={() => import('./pages/root')} />
			<Route path={`/${ROOM_PREFIX}`} lazy={() => import('./pages/new')} />
			<Route path="/new" lazy={() => import('./pages/new')} />
			<Route path={`/ts-side`} lazy={() => import('./pages/public-touchscreen-side-panel')} />
			<Route path={`/${ROOM_PREFIX}/:roomId`} lazy={() => import('./pages/public-multiplayer')} />
			<Route path={`/${ROOM_PREFIX}/:boardId/history`} lazy={() => import('./pages/history')} />
			<Route
				path={`/${ROOM_PREFIX}/:boardId/history/:timestamp`}
				lazy={() => import('./pages/history-snapshot')}
			/>
			<Route path={`/${SNAPSHOT_PREFIX}/:roomId`} lazy={() => import('./pages/public-snapshot')} />
			<Route
				path={`/${READ_ONLY_LEGACY_PREFIX}/:roomId`}
				lazy={() => import('./pages/public-readonly-legacy')}
			/>
			<Route path={`/${READ_ONLY_PREFIX}/:roomId`} lazy={() => import('./pages/public-readonly')} />
		</Route>
		{/* begin tla */}
		<Route lazy={() => import('./tla/providers/TlaClerkProvider')}>
			<Route path="/q/local" lazy={() => import('./tla/pages/local')} />
			<Route lazy={() => import('./tla/providers/TlaAppLoggedInProvider')}>
				{/* If not redirected, then local */}
				<Route path="/q" lazy={() => import('./tla/components/RedirectAtRoot')} />
				{/* File view*/}
				<Route path="/q/f/:fileSlug" lazy={() => import('./tla/pages/file')} />
				{/* User settings */}
				<Route path="/q/profile" lazy={() => import('./tla/components/RequireAuthForUser')}>
					<Route index lazy={() => import('./tla/pages/profile')} />
				</Route>
				{/* Internal */}
				<Route path="/q/debug" lazy={() => import('./tla/pages/debug')} />
			</Route>
		</Route>
		{/* end tla */}
		<Route path="*" lazy={() => import('./pages/not-found')} />
	</Route>
)
