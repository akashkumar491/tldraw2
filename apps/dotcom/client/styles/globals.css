@import url('https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&display=swap');
@import url('tldraw/tldraw.css');
@import url('./z-board.css');

/* NB: If we did position: fixed, Safari on iOS 18 and MacOS 15 would
 * have several rendering issues. This must stay position: absolute. 
 */
.tldraw__editor {
	position: absolute;
	inset: 0;
	width: 100%;
	height: 100%;
	overflow: hidden;
}
