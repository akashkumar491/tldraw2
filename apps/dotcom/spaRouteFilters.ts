// This value is used to configure Vercel routing to rewrite dotcom SPA routes to /index.html.
// It is also tested in routes.test.ts to make sure it matches all React Router routes.
//
// It is a list of string-encoded regexes matching SPA routes to be spliced into
// Vercel's "build output spec" in scripts/build.ts.
//
// Make sure it's not overly broad, because otherwise we won't give correct 404 responses.
export const SPA_ROUTE_FILTERS = ['^/$', '^/r/?$', '^/new/?$', '^/r/.*', '^/s/.*', '^/v/.*']
