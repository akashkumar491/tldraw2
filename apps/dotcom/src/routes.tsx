import { captureException } from '@sentry/react'
import { nanoid } from 'nanoid'
import { useEffect } from 'react'
import { createRoutesFromElements, Outlet, redirect, Route, useRouteError } from 'react-router-dom'
import { DefaultErrorFallback } from './components/DefaultErrorFallback/DefaultErrorFallback'
import { ErrorPage } from './components/ErrorPage/ErrorPage'

export const router = createRoutesFromElements(
	<Route
		element={
			// Add all top level providers that require the router here

			<Outlet />
		}
		ErrorBoundary={() => {
			const error = useRouteError()
			useEffect(() => {
				captureException(error)
			}, [error])
			return (
				<ErrorPage
					messages={{
						header: 'Something went wrong',
						para1:
							'Please try refreshing the page. Still having trouble? Let us know at hello@tldraw.com.',
					}}
				/>
			)
		}}
	>
		<Route errorElement={<DefaultErrorFallback />}>
			<Route path="/" lazy={() => import('./pages/root')} />
			<Route
				path="/r"
				loader={() => {
					const id = 'v2' + nanoid()
					return redirect(`/r/${id}`)
				}}
			/>
			<Route
				path="/new"
				loader={() => {
					const id = 'v2' + nanoid()
					return redirect(`/r/${id}`)
				}}
			/>
			<Route path="/r/:roomId" lazy={() => import('./pages/public-multiplayer')} />
			<Route path="/r/:boardId/history" lazy={() => import('./pages/history')} />
			<Route
				path="/r/:boardId/history/:timestamp"
				lazy={() => import('./pages/history-snapshot')}
			/>
			<Route path="/s/:roomId" lazy={() => import('./pages/public-snapshot')} />
			<Route path="/v/:roomId" lazy={() => import('./pages/public-readonly')} />
		</Route>
		<Route path="*" lazy={() => import('./pages/not-found')} />
	</Route>
)
