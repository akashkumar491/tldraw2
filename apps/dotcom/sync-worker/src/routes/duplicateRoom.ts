import { TldrawAppFileRecordType } from '@tldraw/dotcom-shared'
import { uniqueId } from '@tldraw/utils'
import { IRequest } from 'itty-router'
import { Environment } from '../types'
import { isRoomIdTooLong } from '../utils/roomIdIsTooLong'
import { getRoomCurrentSnapshot } from '../utils/tla/getRoomCurrentSnapshot'
import { getTldrawAppDurableObject } from '../utils/tla/getTldrawAppDurableObject'
import {
	fileOwnerStatusErrorResponse,
	getFileOwnerStatus,
	getUserIdFromRequest,
} from '../utils/tla/permissions'
import { updateRoomSnapshot } from '../utils/tla/updateRoomSnapshot'

export async function duplicateRoom(request: IRequest, env: Environment): Promise<Response> {
	const { roomId } = request.params

	const userId = await getUserIdFromRequest(request, env)
	if (!userId) {
		return Response.json({ error: true, message: 'No user' }, { status: 401 })
	}

	const status = await getFileOwnerStatus(request, env, roomId)
	if (!status.ok) {
		return fileOwnerStatusErrorResponse(status.error)
	}

	if (!roomId) {
		return Response.json({ error: true, message: 'Room ID is required' }, { status: 400 })
	}

	if (isRoomIdTooLong(roomId)) {
		return Response.json({ error: true, message: 'Room ID is too long' }, { status: 400 })
	}

	// Get the current serialized snapshot of the room (by waking up the worker,
	// if we have to). Why not grab from the database directly? Because the worker
	// only persists every ~30s while users are actively editing the room. If we
	// knew whether the room was active or not (either by checking whether the
	// worker was awake or somehow recording which rooms have active users in them,
	// or when the room was last edited) we could make a better decision.
	const serializedSnapshot = await getRoomCurrentSnapshot(roomId, env)

	// Create a new slug for the duplicated room
	const slug = uniqueId()

	// Bang the snapshot into the database under a different slug
	await updateRoomSnapshot(slug, serializedSnapshot, env)

	// Now create a new file in the app durable object
	const appDo = getTldrawAppDurableObject(env)
	const fileId = TldrawAppFileRecordType.createId(slug)
	await appDo.createNewFile(
		TldrawAppFileRecordType.create({
			id: fileId,
			ownerId: userId,
		})
	)

	// Send back the slug so that the client can redirect to the new room
	return new Response(JSON.stringify({ error: false, slug }))
}
