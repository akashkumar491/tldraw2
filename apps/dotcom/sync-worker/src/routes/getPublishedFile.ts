import { RoomSnapshot } from '@tldraw/sync-core'
import { IRequest } from 'itty-router'
import { Environment } from '../types'
import { getTldrawAppDurableObject } from '../utils/tla/getTldrawAppDurableObject'

export async function getPublishedFile(request: IRequest, env: Environment): Promise<Response> {
	const { roomId } = request.params
	if (!roomId) {
		return Response.json({ error: true, message: 'Room ID is required' }, { status: 400 })
	}

	const app = getTldrawAppDurableObject(env)

	try {
		const publishedRoomSnapshot = await (app.getPublishedFile(roomId) as Promise<RoomSnapshot>)
		const { documents, schema } = publishedRoomSnapshot
		return new Response(
			JSON.stringify({
				records: documents.map((d) => d.state),
				schema: schema,
				error: false,
			}),
			{
				headers: { 'content-type': 'application/json' },
			}
		)
	} catch (e: any) {
		return new Response(JSON.stringify({ error: true, message: e.message }), { status: 500 })
	}
}
