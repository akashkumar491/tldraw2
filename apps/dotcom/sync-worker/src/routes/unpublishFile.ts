import { IRequest } from 'itty-router'
import { Environment } from '../types'
import { getTldrawAppDurableObject } from '../utils/tla/getTldrawAppDurableObject'
import { getUserIdFromRequest } from '../utils/tla/permissions'

export async function unpublishFile(request: IRequest, env: Environment): Promise<Response> {
	const { roomId } = request.params
	if (!roomId) {
		return Response.json({ error: true, message: 'Room ID is required' }, { status: 400 })
	}

	const userId = await getUserIdFromRequest(request, env)
	if (!userId) {
		return Response.json({ error: true, message: 'No user' }, { status: 401 })
	}

	const app = getTldrawAppDurableObject(env)

	try {
		await app.unpublishFile(roomId, userId)
		return new Response(JSON.stringify({ error: false }))
	} catch (e: any) {
		return new Response(JSON.stringify({ error: true, message: e.message }), { status: 500 })
	}
}
