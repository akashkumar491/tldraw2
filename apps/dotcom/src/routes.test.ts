import { expect } from '@jest/globals'
import type { MatcherFunction } from 'expect'
import { RouteObject } from 'react-router-dom'
import { SPA_ROUTE_FILTERS } from '../spaRouteFilters'
import { router } from './routes'

const toMatchAny: MatcherFunction<[regexes: unknown]> = function (actual, regexes) {
	if (
		typeof actual !== 'string' ||
		!Array.isArray(regexes) ||
		regexes.some((regex) => typeof regex !== 'string')
	) {
		throw new Error('toMatchAny takes a string and an array of strings')
	}

	const pass = regexes.map((regex) => new RegExp(regex)).some((regex) => actual.match(regex))
	if (pass) {
		return {
			message: () =>
				`expected ${this.utils.printReceived(actual)} not to match any of the regexes ${this.utils.printExpected(regexes)}`,
			pass: true,
		}
	} else {
		return {
			message: () =>
				`expected ${this.utils.printReceived(actual)} to match at least one of the regexes ${this.utils.printExpected(regexes)}`,
			pass: false,
		}
	}
}

expect.extend({ toMatchAny })

function extractContentPaths(routeObject: RouteObject): string[] {
	const paths: string[] = []

	if (routeObject.path && routeObject.path !== '*') {
		paths.push(routeObject.path)
	}

	if (routeObject.children) {
		routeObject.children.forEach((child) => {
			paths.push(...extractContentPaths(child))
		})
	}

	return paths
}

test('SPA_ROUTE_FILTERS match all React routes', () => {
	router.flatMap(extractContentPaths).forEach((path) => {
		expect(path).toMatchAny(SPA_ROUTE_FILTERS)
	})
})

test("SPA_ROUTE_FILTERS don't match assets", () => {
	expect('/assets/test.png').not.toMatchAny(SPA_ROUTE_FILTERS)
})
