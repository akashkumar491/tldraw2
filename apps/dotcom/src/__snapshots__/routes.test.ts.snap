// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the_routes 1`] = `
[
  {
    "reactRouterPattern": "/",
    "vercelRoutingPattern": "^//?$",
  },
  {
    "reactRouterPattern": "/new",
    "vercelRoutingPattern": "^/new/?$",
  },
  {
    "reactRouterPattern": "/r",
    "vercelRoutingPattern": "^/r/?$",
  },
  {
    "reactRouterPattern": "/r/:boardId/history",
    "vercelRoutingPattern": "^/r/[^/]*/history/?$",
  },
  {
    "reactRouterPattern": "/r/:boardId/history/:timestamp",
    "vercelRoutingPattern": "^/r/[^/]*/history/[^/]*/?$",
  },
  {
    "reactRouterPattern": "/r/:roomId",
    "vercelRoutingPattern": "^/r/[^/]*/?$",
  },
  {
    "reactRouterPattern": "/s/:roomId",
    "vercelRoutingPattern": "^/s/[^/]*/?$",
  },
  {
    "reactRouterPattern": "/sistine/chapel",
    "vercelRoutingPattern": "^/sistine/chapel/?$",
  },
  {
    "reactRouterPattern": "/v/:roomId",
    "vercelRoutingPattern": "^/v/[^/]*/?$",
  },
]
`;
